b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+  _port: -1, // The port on which this service listens
+  _socket: null, // The socket associated with this
+  _doQuit: false, // Indicates when the service is to be shut down at the end of the request.
+  _socketClosed: true, // True if the socket in this is closed, false otherwise.
+  _connectionGen: 0, // Used for tracking existing connections
+  _connections: new Map(), // Hash of all open connections, indexed by connection number
+  _mDNSCancelableHandler: null, // Handle mDNS cancel
+
+  init: function() {
+    DEBUG && debug("init");
+  },
+
+  // PUBLIC API
+  // Start TLS socket server.
+  // Return a promise for start() resolves/reject to
+  start: function() {
+    if (this._serverStatus == SERVER_STATUS.STARTED) {
+      return Promise.reject("AlreadyStarted");
+    }
+
+    let promise = new Promise((aResolve, aReject) => {
+      this._doStart(aResolve, aReject);
+    });
+    return promise;
+  },
+
+  // Stop TLS socket server, remove registered observer
+  // Cancel mDNS registration
+  // Return false if server not started, stop failed.
+  stop: function() {
+    if (this._serverStatus == SERVER_STATUS.STOPPED) {
+      return false;
+    }
+
+    if (!this._socket) {
+      return false;
+    }
+
+    DEBUG && debug("Stop listening on port " + this._socket.port);
+
+    Services.obs.removeObserver(this, "xpcom-shutdown");
+
+    if (this._mDNSCancelableHandler) {
+      this._mDNSCancelableHandler.Cancel(Cr.NS_OK);
+      this._mDNSCancelableHandler = null;
+    }
+
+    this._socket.close();
+    this._socket = null;
+    this._doQuit = false;
+    this._serverStatus = SERVER_STATUS.STOPPED;
+
+    return true;
+  },
+
+  // Observers and Listeners
+  // nsIObserver
+  observe: function(subject, topic, data) {
+    switch (topic) {
+      case "xpcom-shutdown": {
+        // Stop service when xpcom-shutdown
+        this.stop();
+        break;
+      }
+    }
+  },
+
+  // nsIServerSocketListener
+  onSocketAccepted: function(socket, trans) {
+    DEBUG && debug("onSocketAccepted(socket=" + socket + ", trans=" + trans + ")");
+    DEBUG && debug("New connection on " + trans.host + ":" + trans.port);
+
+    const SEGMENT_SIZE = 8192;
+    const SEGMENT_COUNT = 1024;
+
+    try {
+      var input = trans.openInputStream(0, SEGMENT_SIZE, SEGMENT_COUNT)
+                       .QueryInterface(Ci.nsIAsyncInputStream);
+      var output = trans.openOutputStream(0, 0, 0);
+    } catch (e) {
+      DEBUG && debug("Error opening transport streams: " + e);
+      trans.close(Cr.NS_BINDING_ABORTED);
+      return;
+    }
+
+    let connectionNumber = ++this._connectionGen;
+
+    try {
+      // Create a connection for each user connection
+      // EventHandler implements nsIInputStreamCallback for incoming message from user
+      var conn = new Connection(input, output, this, socket.port, trans.port, connectionNumber);
+      let handler = new EventHandler(conn);
+
+      input.asyncWait(handler, 0, 0, Services.tm.mainThread);
```

Comment:
You should combine Connection and EventHandler into one class, or at least move the I/O stream handling into Connection. In addition, you should also move `input.asyncWait` into Connection.